# 1. Загрузка и подготовка данных

In [1]:
#импорт необходимых библиотек
import pandas as pd 
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from numpy.random import RandomState
import matplotlib.pyplot as plt
from scipy.stats import skewnorm

In [2]:
#загрузка базы
data_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_2 = pd.read_csv('/datasets/geo_data_2.csv')

In [3]:
#проверка выдачи по базам
display(data_0.head(10))
display(data_1.head(10))
display(data_2.head(10))

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
5  wX4Hy  0.969570  0.489775 -0.735383   64.741541
6  tL6pL  0.645075  0.530656  1.780266   49.055285
7  BYPU6 -0.400648  0.808337 -5.624670   72.943292
8  j9Oui  0.643105 -0.551583  2.372141  113.356160
9  OLuZU  2.173381  0.563698  9.441852  127.910945

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
5  HHckp  -3.327590  -2.205276  3.003647   84.038886
6  h5Ujo -11.142655 -10.133399  4.002382  110.992147
7  muH9x   4.234715  -0.001354  2.004588   53.906522
8  YiRkx  13.355129  -0.332068  4.998647  134.766305
9  jG6Gi   1.069227 -11.025667  4.997844  137.945408

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
5  LzZXx -0.758092  0.710691  2.585887   90.222465
6  WBHRv -0.574891  0.317727  1.773745   45.641478
7  XO8fn -1.906649 -2.458350 -0.177097   72.480640
8  ybmQ5  1.776292 -0.279356  3.004156  106.616832
9  OilcN -1.214452 -0.439314  5.922514   52.954532

In [4]:
#удаляем лишние столбцы из данных
data_0 = data_0.drop('id', axis=1)
data_1 = data_1.drop('id', axis=1)
data_2 = data_2.drop('id', axis=1)

In [5]:
#просмотр информации по базам
data_0.info()
data_1.info()
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4)
memory usage: 3.1 MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4)
memory usage: 3.1 MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4)
memory usage: 3.1 MB


# Шаг 1. Выводы   
1) Извлекли нужные библиотеки для дальнейшей работы.   
2) Загрузили баззы данных.  
3) Проверили выдачу по базам и посомтрели информацию по базам (наличие пропущенных значений) Базы полностью заполнены, пропущенных значений нет. 

# 2. Обучение и проверка модели

In [6]:
#разделение на обучающую и тестовую выборку (data_0)
data_train_0, data_valid_0 = train_test_split(data_0, test_size=0.25, random_state=12345)
features_train_0 = data_train_0.drop(['product'], axis=1)
target_train_0 = data_train_0['product']
features_valid_0 = data_valid_0.drop(['product'], axis=1)
target_valid_0 = data_valid_0['product']

In [7]:
#разделение на обучающую и тестовую выборку (data_1)
data_train_1, data_valid_1 = train_test_split(data_1, test_size=0.25, random_state=12345)
features_train_1 = data_train_1.drop(['product'], axis=1)
target_train_1 = data_train_1['product']
features_valid_1 = data_valid_1.drop(['product'], axis=1)
target_valid_1 = data_valid_1['product']

In [8]:
#разделение на обучающую и тестовую выборку (data_2)
data_train_2, data_valid_2 = train_test_split(data_2, test_size=0.25, random_state=12345)
features_train_2 = data_train_2.drop(['product'], axis=1)
target_train_2 = data_train_2['product']
features_valid_2 = data_valid_2.drop(['product'], axis=1)
target_valid_2 = data_valid_2['product']

In [9]:
#обучение модели
model_0 = LinearRegression()
model_1 = LinearRegression()
model_2 = LinearRegression()
model_0.fit(features_train_0, target_train_0)
model_1.fit(features_train_1, target_train_1)
model_2.fit(features_train_2, target_train_2)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [10]:
#предсказание модели
predicted_0 = model_0.predict(features_valid_0)
predicted_1 = model_1.predict(features_valid_1)
predicted_2 = model_2.predict(features_valid_2)

In [11]:
#средний запас сырья на предсказанных данных по регионам
print('Средний запас предсказанного сырья по валидационным данным из data_0:', predicted_0.mean())
print('Средний запас предсказанного сырья по валидационным данным из data_1:', predicted_1.mean())
print('Средний запас предсказанного сырья по валидационным данным из data_2:', predicted_2.mean())

Средний запас предсказанного сырья по валидационным данным из data_0: 92.59256778438038
Средний запас предсказанного сырья по валидационным данным из data_1: 68.728546895446
Средний запас предсказанного сырья по валидационным данным из data_2: 94.96504596800489


In [12]:
#RMSE по регионам
#predictions_0 = pd.Series(target_valid_0.mean(), index=target_valid_0.index)
print('RMSE по валидационным данным из data_0:', mean_squared_error(target_valid_0, predicted_0)**0.5)

#predictions_1 = pd.Series(target_valid_1.mean(), index=target_valid_1.index)
print('RMSE по валидационным данным из data_1:', mean_squared_error(target_valid_1, predicted_1)**0.5)

#predictions_2 = pd.Series(target_valid_2.mean(), index=target_valid_2.index)
print('RMSE по валидационным данным из data_2:', mean_squared_error(target_valid_2, predicted_2)**0.5)



RMSE по валидационным данным из data_0: 37.5794217150813
RMSE по валидационным данным из data_1: 0.893099286775616
RMSE по валидационным данным из data_2: 40.02970873393434


# Шаг 2. Выводы   
1) Проведено разделение на обучающую и валидационную выборки каждую базу по региону.   
2) Проведено обучение модели - линейная регресиия (из описания проекта сказано, что она лучше всех)   
3) Предсказаны данные по каждой валиадционной выборки   
4) Найдено среднее значение по запасам на предсказанных данных.  
5) Расчитано RMSE по каждой базе. Данные по базе 0 и 2 сильно больше, чем у базы 1. Значение по базе 1 почти идельное (чуть-чуть не хватает)

# 3. Подготовка к расчёту прибыли

In [13]:
#бюджет выделяемый на разработку скважин в регион
BUDGET = 10000000000

#стоимость 1 барреля 
COST_BAR = 450000

#кол-во скважин для разработки в каждом региона
WELL = 200

#находим объем сырья для безубыточной разработки
min_bar_reg = BUDGET / COST_BAR

#находим объем сырья для 200 лучших скважин из каждого региона для безубыточной разработки 
min_bar_well = min_bar_reg / WELL


print('Минимальный объем сырья для безубыточной разработки в регионе:', min_bar_reg)
print('Минимальный объем сырья для каждой скважины для безубыточной разработки в регионе:', min_bar_well)


Минимальный объем сырья для безубыточной разработки в регионе: 22222.222222222223
Минимальный объем сырья для каждой скважины для безубыточной разработки в регионе: 111.11111111111111


In [14]:
#средний запас сырья на предсказанных данных по регионам
print('Средний запас предсказанного сырья по валидационным данным из data_0:', predicted_0.mean())
print('Средний запас предсказанного сырья по валидационным данным из data_1:', predicted_1.mean())
print('Средний запас предсказанного сырья по валидационным данным из data_2:', predicted_2.mean())

Средний запас предсказанного сырья по валидационным данным из data_0: 92.59256778438038
Средний запас предсказанного сырья по валидационным данным из data_1: 68.728546895446
Средний запас предсказанного сырья по валидационным данным из data_2: 94.96504596800489


In [15]:
#разница между минимальным объемом сырья для безубыточной разработки (МОСДБР) и средним объем запаса сырья из предсказаных данных по регионам
difference_data_0 = min_bar_well - predicted_0.mean()
difference_data_1 = min_bar_well - predicted_1.mean()
difference_data_2 = min_bar_well - predicted_2.mean()

print('Разница между минимальным МОСДБР и средним объем сырья из данных data_0:', difference_data_0)
print('Разница между минимальным МОСДБР и средним объем сырья из данных data_1:', difference_data_1)
print('Разница между минимальным МОСДБР и средним объем сырья из данных data_2:', difference_data_2)

Разница между минимальным МОСДБР и средним объем сырья из данных data_0: 18.518543326730736
Разница между минимальным МОСДБР и средним объем сырья из данных data_1: 42.38256421566511
Разница между минимальным МОСДБР и средним объем сырья из данных data_2: 16.146065143106227


# Шаг 3. Выводы   
1) Найден минимальный объем сырья, который сможет обеспечить нам безубыточность в регионе. - 22,2к  
2) Найден минимальный объем сырья для каждой разрабатываемой скважины в регионе.  -  111,1.    
3) Найденный минимальный объем сырья больше, чем средений запас сырья в каждом регионе. Разница от 16.1 до 42.3 в зависимости от региона. (данные по data_0 и data_2 - рядом 16 и 18, сильно вырывается вперед данные по data_1 - 42.3)

# 4. Функция для расчета прибыли


In [16]:
#соединяю данные предсказыне и целевые в каждом регионе 
#data_0
target_valid_0 = data_valid_0.reset_index().drop('index', axis=1)
predicted_0 = pd.DataFrame(predicted_0)
predicted_0.columns = ['product_pred']
pred_target_valid_0 = pd.concat((target_valid_0, predicted_0), axis=1)
print(pred_target_valid_0.head(10))

#data_1
target_valid_1 = data_valid_1.reset_index().drop('index', axis=1)
predicted_1 = pd.DataFrame(predicted_1)
predicted_1.columns = ['product_pred']
pred_target_valid_1 = pd.concat((target_valid_1, predicted_1), axis=1)
print(pred_target_valid_1.head(10))

#data_2
target_valid_2 = data_valid_2.reset_index().drop('index', axis=1)
predicted_2 = pd.DataFrame(predicted_2)
predicted_2.columns = ['product_pred']
pred_target_valid_2 = pd.concat((target_valid_2, predicted_2), axis=1)
print(pred_target_valid_2.head(10))

         f0        f1        f2     product  product_pred
0  0.948970 -0.057547  2.095727   10.038645     95.894952
1  0.992974  0.206671 -0.142278  114.551489     77.572583
2  1.199854 -0.563356 -1.852991  132.603635     77.892640
3  0.691422 -0.433347  0.564974  169.072125     90.175134
4  0.420772  0.972638  0.736190  122.325180     70.510088
5 -0.502687  0.729647  0.510063   37.735331     69.127076
6  1.358998 -0.569837  5.207498  139.227191    125.106759
7  0.949783  0.187695  1.368214   72.491987     87.643849
8  0.742510  0.733056  2.403405   11.465993     86.035871
9  1.911387  0.023812  2.163916  110.562031     98.655311
          f0         f1        f2     product  product_pred
0  -0.371866  -1.862494  3.002210   80.859783     82.663314
1   9.015122 -13.881455  1.995363   53.906522     54.431786
2  -6.507568  -4.817448  1.003449   30.132364     29.748760
3  14.560845 -10.667755  1.995175   53.906522     53.552133
4   6.090476  -4.494723  0.013815    0.000000      1.243856
5 

In [17]:
#функция подсчета прибыли
def profit_top_well(data):
    data = data.sort_values(by = 'product_pred', ascending = False)
    data = data[:WELL]
    profit_bar = round(data['product'].sum() *  COST_BAR / 1000000000 - 10, 2)
    return profit_bar

In [18]:
#подсчет прибыли с каждого региона

print("Прибыль с топ-200 скважин по предсказанным данным", profit_top_well(pred_target_valid_0), 'млрд рублей')
print("Прибыль с топ-200 скважин по предсказанным данным", profit_top_well(pred_target_valid_1), 'млрд рублей')
print("Прибыль с топ-200 скважин по предсказанным данным", profit_top_well(pred_target_valid_2), 'млрд рублей')

Прибыль с топ-200 скважин по предсказанным данным 3.32 млрд рублей
Прибыль с топ-200 скважин по предсказанным данным 2.42 млрд рублей
Прибыль с топ-200 скважин по предсказанным данным 2.71 млрд рублей


# Шаг 4. Выводы   
По топ-200 по предсказанным данным (но подсчеты проводим по целевым значениям), добыча нефти в каждом регионе будет прибыльной, так как план по доходу получается более 12 млрд рублей - при затратах в 10 млрд рублей.   
Итого профит: от 2,4 млрд до 3,3 млрд


# 5. Подсчеты рисков и прибыли для каждого региона

In [19]:
#создаем 1000 выборок при помощи бутстрепа для каждого региона (на предсказанных данных) и находим среднее по средним
state = np.random.RandomState(12345)

#data_0
sum_profit_0 = 0
for i in range(1000):
    subsample = pred_target_valid_0.sample(n=500, replace=True, random_state=state)
    sum_profit_0 += profit_top_well(subsample)
mean_profit_0 = round((sum_profit_0 / 1000), 2)
print('Среднее по выборочным средним из топ-200 каждой выборки из данных data_0:', mean_profit_0, 'млрд рублей')

#data_1
sum_profit_1 = 0
for i in range(1000):
    subsample = pred_target_valid_1.sample(n=500, replace=True, random_state=state)
    sum_profit_1 += profit_top_well(subsample)
mean_profit_1 = round((sum_profit_1 / 1000), 2)
print('Среднее по выборочным средним из топ-200 каждой выборки из данных data_1:', mean_profit_1, 'млрд рублей')

#data_2
sum_profit_2 = 0
for i in range(1000):
    subsample = pred_target_valid_2.sample(n=500, replace=True, random_state=state)
    sum_profit_2 += profit_top_well(subsample)
mean_profit_2 = round((sum_profit_2 / 1000), 2)
print('Среднее по выборочным средним из топ-200 каждой выборки из данных data_2:', mean_profit_2, 'млрд рублей')

Среднее по выборочным средним из топ-200 каждой выборки из данных data_0: 0.4 млрд рублей
Среднее по выборочным средним из топ-200 каждой выборки из данных data_1: 0.46 млрд рублей
Среднее по выборочным средним из топ-200 каждой выборки из данных data_2: 0.39 млрд рублей


In [20]:
#находим 95% интервал
pred_target = [pred_target_valid_0, pred_target_valid_1, pred_target_valid_2]
for element in pred_target:
    mean_profit_top_well = []
    minus = 0
    for i in range(1000):
        subsample = element.sample(n=500, replace=True, random_state=state)
        mean_profit_top_well.append(profit_top_well(subsample))
        if mean_profit_top_well[i] < 0:
            minus += 1

    mean_profit_top_well = pd.Series(mean_profit_top_well)

    lower = mean_profit_top_well.quantile(0.025)
    upper = mean_profit_top_well.quantile(0.975)
    
    print('95% доверительный интервал на данных')
    print(lower)
    print(upper)
    
#находим риск убытков    
    print('Процент убыточных топовых скважин:', minus / 1000 * 100)
    print('--')

#находим среднее по топовым скважинам
    print('Среднее по топовым скважинам:', mean_profit_top_well.mean())
    print('--------------------')


95% доверительный интервал на данных
-0.12024999999999997
0.91
Процент убыточных топовых скважин: 5.6000000000000005
--
Среднее по топовым скважинам: 0.39370000000000005
--------------------
95% доверительный интервал на данных
0.05
0.84
Процент убыточных топовых скважин: 1.2
--
Среднее по топовым скважинам: 0.44817
--------------------
95% доверительный интервал на данных
-0.2
0.93
Процент убыточных топовых скважин: 9.0
--
Среднее по топовым скважинам: 0.38258000000000003
--------------------


In [25]:
#находим 95% интервал
#data_0
mean_profit_top_well_0 = []
for i in range(1000):
    subsample = pred_target_valid_0.sample(n=500, replace=True, random_state=state)
    mean_profit_top_well_0.append(profit_top_well(subsample))

mean_profit_top_well_0 = pd.Series(mean_profit_top_well_0)
    
lower_0 = mean_profit_top_well_0.quantile(0.025)
upper_0 = mean_profit_top_well_0.quantile(0.975)

print('95% доверительный интервал на данных data_0')
print(lower_0)
print(upper_0)
print('-----')

#находим риск убытков
minus_0 = 0
for i in range(1000):
    if mean_profit_top_well_0[i] < 0:
        minus_0 += 1
unprofitable_wells_0 = minus_0 / 1000 * 100
print('Процент убыточных топовых скважин:', unprofitable_wells_0)
print('-----')

#находим среднее по топовым скважинам
print('Среднее по топовым скважинам из data_0:', mean_profit_top_well_0.mean())


95% доверительный интервал на данных data_0
-0.13
0.8902499999999998
-----
Процент убыточных топовых скважин: 6.4
-----
Среднее по топовым скважинам из data_0: 0.39086000000000004


In [26]:
#находим 95% интервал
#data_1
mean_profit_top_well_1 = []
for i in range(1000):
    subsample = pred_target_valid_1.sample(n=500, replace=True, random_state=state)
    mean_profit_top_well_1.append(profit_top_well(subsample))

mean_profit_top_well_1 = pd.Series(mean_profit_top_well_1)
    
lower_1 = mean_profit_top_well_1.quantile(0.025)
upper_1 = mean_profit_top_well_1.quantile(0.975)

print('95% доверительный интервал на данных data_1')
print(lower_1)
print(upper_1)
print('-----')

#находим риск убытков
minus_1 = 0
for i in range(1000):
    if mean_profit_top_well_1[i] < 0:
        minus_1 += 1
unprofitable_wells_1 = minus_1 / 1000 * 100
print('Процент убыточных топовых скважин:', unprofitable_wells_1)
print('-----')

#находим среднее по топовым скважинам
print('Среднее по топовым скважинам из data_1:', mean_profit_top_well_1.mean())

95% доверительный интервал на данных data_1
0.06975000000000002
0.85
-----
Процент убыточных топовых скважин: 0.8
-----
Среднее по топовым скважинам из data_1: 0.45363


In [27]:
#находим 95% интервал
#data_2
mean_profit_top_well_2 = []
for i in range(1000):
    subsample = pred_target_valid_2.sample(n=500, replace=True, random_state=state)
    mean_profit_top_well_2.append(profit_top_well(subsample))

mean_profit_top_well_2 = pd.Series(mean_profit_top_well_2)
    
lower_2 = mean_profit_top_well_2.quantile(0.025)
upper_2 = mean_profit_top_well_2.quantile(0.975)

print('95% доверительный интервал на данных data_2')
print(lower_2)
print(upper_2)
print('-----')

#находим риск убытков
minus_2 = 0
for i in range(1000):
    if mean_profit_top_well_2[i] < 0:
        minus_2 += 1
unprofitable_wells_2 = minus_2 / 1000 * 100
print('Процент убыточных топовых скважин:', unprofitable_wells_2)
print('-----')

#находим среднее по топовым скважинам
print('Среднее по топовым скважинам из data_2:', mean_profit_top_well_2.mean())

95% доверительный интервал на данных data_2
-0.11
0.94
-----
Процент убыточных топовых скважин: 6.6000000000000005
-----
Среднее по топовым скважинам из data_2: 0.39997000000000005


In [28]:
#таблица с общими результатами
well_data = ['data_0', 'data_1', 'data_2']
data = {
  '95% доверительный интервал_min': [lower_0, lower_1, lower_2],
  '95% доверительный интервал_max': [upper_0, upper_1, upper_2],
  'Процент убыточных топовых скважин':  [unprofitable_wells_0, unprofitable_wells_1, unprofitable_wells_2],
  'Среднее по топовым скважинам':  [mean_profit_top_well_0.mean(), mean_profit_top_well_1.mean(), mean_profit_top_well_2.mean()]
}
data_ml = pd.DataFrame(data=data, index=well_data)
display(data_ml)

95% доверительный интервал_min  95% доверительный интервал_max  \
data_0                        -0.13000                         0.89025   
data_1                         0.06975                         0.85000   
data_2                        -0.11000                         0.94000   

        Процент убыточных топовых скважин  Среднее по топовым скважинам  
data_0                                6.4                       0.39086  
data_1                                0.8                       0.45363  
data_2                                6.6                       0.39997

# Шаг 5. Выводы   
  
1) Самый главный вывод проделанной работы, что разрабатывать нужно скважины из геопозиции - data_1   
а) Во-первых - у этих данных самый низкий процент убыточных скважин   
б) Во-вторых - у этих данных получился самый лучший доверительный интервал   
в) В-третьих - у этих данных самый высокий показатель среднего из топовых скважин